README.md
# measly

> A measly wrapper around XHR to help you contain your requests

# Online Demo

[![measly.png][3]][4]

# Install

```shell
npm install measly --save
```

```shell
bower install measly --save
```

# Layer API

Here is the API exposed by `measly`.

## `.layer(options?)`

Measly works as a hierarchy-based layer that helps you perform XHR requests. `measly.layer()` provides the same API as `measly` does, by creating a child layer. Measly supports tree-like structures as well. You can use a thinner layer to create requests from different parts of your application. You can pass a few options to `measly.layer`.

Option    | Description
----------|--------------------------------------------------------------------------------------------
`context` | A DOM context element for the layer. Defaults to `document.body`.
`cache`   | Milliseconds that response data is considered fresh. Defaults to `false`.

Note that if you pass in a `context` that already has a Measly layer, that layer will be returned instead.

#### Usage

```js
var core = require('measly');
var thin = core.layer();

// thin has same api as core
```

## `.parent`

Access the parent layer.

#### Usage

```js
var core = require('measly');
var thin = core.layer();

console.log(core.parent);
// <- undefined

console.log(thin.parent);
// <- core
```

## `.context`

This is the context element for the measly layer.

#### Usage

```js
var core = require('measly');
var thin = core.layer({ context: div });

console.log(core.context);
// <- document.body

console.log(thin.context);
// <- div
```

## `.all`

Lists all of the measly layers that have been ever created. **Note that this method is only available in the top measly layer.**

## `.find(context, shallow?)`

Finds a measly layer by their `context` DOM element. If the provided element isn't found in any layer, its parent element is looked up. If no other parent elements are found, then you're going to get `core` back because its context is `document.body`. **Note that this method is only available in the top measly layer.**

If you set `shallow` to `true`, then only the provided `context` will be tested, rather than walking the DOM tree.

#### Usage

```js
var core = require('measly');
var thin = core.layer({ context: div });

console.log(core.find(document.body));
// <- core

console.log(core.find(div));
// <- thin
```

## `.children`

Array of thinner layers created on top of this layer.

#### Usage

```js
var core = require('measly');
var thin = core.layer();

console.log(core.children);
// <- [thin]
```

## `.requests`

These are the pending requests that pertain to this measly layer.

#### Usage

```js
var core = require('measly');
core.get('/api/v1/logs');

console.log(core.requests);
// <- [req]
```

## `.cache`

These are the cache entries for the current layer. Note that freshness is evaluated only when the cache entry is considered as a response, and thus the `cache` object may have stale entries in it, even though they will be removed when detected.

Whenever there's a cache hit the request won't be initiated against the server. Instead, your local copy of the data will be used. **Cache hits are limited to `GET` requests because best practices.**

#### Usage

```js
var core = require('measly');

core.cache['/api/v1/logs'] = {
  statusCode: 200,
  body: 'a cached response'
};

core.get('/api/v1/logs').on('data', function (err, res, body) {
  console.log(body);
  // <- 'a cached response'
});
```

**Note that the `cache` properties are automatically set for you if you define a duration for which your requests are _cache-worthy_.**

```js
var core = require('measly');
core.get('/api/v1/logs', { cache: 60000 }).on('data', function () {
  core.get('/api/v1/logs').on('data', function () {
    // same response. the resource won't be hit again for 60 seconds
  });
});
```

## `.abort()`

This method will abort all pending requests found on this layer as well as all its children's requests, recursively.

#### Usage

```js
var core = require('measly');

core.get('/api/v1/logs');
core.abort();
```

## `.request(url, options)`

Requests are created using [`xhr`][1]. In addition to the options defined by `xhr`, you could define these options as well.

Option    | Description
----------|--------------------------------------------------------------------------------------------
`context` | An arbitrary context object for this. DOM elements are encouraged. Defaults to `layer.context`.
`cache`   | Milliseconds that response data is considered fresh. Defaults to `layer.cache`.

There's also `.get`, `.post`, `.put`, `.delete`, and `.patch` as short-hand methods. These methods return a `req` object. Each instance of `req` has an extensive API.

#### Usage

```js
var core = require('measly');
var req = core.get('/api/v1/logs');
```

# Request API

Firstly, `req` objects are also event emitters created by [`contra`][2]. These events don't only get fired on the `req` object, but on the `layer` that created the request, and all of its parents.

## Request Events

In every event listener, `req` will be the context assigned to `this`.

Event          | Arguments     | Fired when...
---------------|---------------|-----------------------------------------------------------
`'create'`     | `(req)`       | A `measly` request is created
`'cache'`      | `(err, body)` | A request is prevented either manually or by a cache hit
`'request'`    | `(xhr)`       | The XHR request is opened
`'abort'`      | `(xhr?)`      | A request is aborted manually
`'error'`      | `(err, body)` | There was an error
`(statusCode)` | `(err, body)` | There was an error with status code: `statusCode`
`'data'`       | `(body)`      | We got a successful response
`'always'`     | `(err, body)` | A request produces an error, is fulfilled, or aborted

Naturally, requests also have properties and other methods.

## `.prevent(err, body, statusCode?)`

Prevents a request from turning into an XHR request. Instead, you can define your own error, response body, and status code. Considering events get bubbled up `measly` layers, you're able to prevent requests under certain conditions, globally for any requests made by a layer or its children. Also useful in testing environments.

The status code is `200` if there isn't an error, `500` in case of an error, and you can also set it explicitly.

#### Usage

```js
var core = require('measly');
var req = core.get('/api/v1/logs');

core.on('create', function (req) {
  if (canInfer) {
    req.prevent(null, ['dogs', 'cats', 'carrots']);
  }
});
```

## `.abort()`

Aborts the request individually.

## `.layer`

This is the layer the request was created on.

#### Usage

```js
var core = require('measly');
var req = core.get('/api/v1/logs');

console.log(req.layer);
// <- core
```

## `.context`

An arbitrary context object for this. DOM elements are encouraged. Defaults to `layer.context`.

## `.cache`

Milliseconds that response data is considered fresh. Defaults to `layer.cache`.

## `.done`

`false` unless the request is completed.

## `.requested`

`false` unless an XHR request is actually created.

## `.prevented`

`false` unless the request is prevented either manually or by a cache hit

## `.xhr`

The browser native XHR object, once the request is created.

# License

MIT

[1]: https://github.com/Raynos/xhr
[2]: https://github.com/bevacqua/contra
[3]: https://cloud.githubusercontent.com/assets/934293/3533872/66285232-07dc-11e4-9601-b7f4ae07bf3e.png
[4]: http://bevacqua.github.io/measly/


CHANGELOG.md
# 1.3.8 Aural Emission

- Replaced `contra.emitter` with `contra@1.9.1/emitter`

# 1.3.6 Mayday Mayday

- Introduced `.aborted` property on individual requests

# 1.3.5 Carl Marx

- Removed dependency on `lodash.find`

# 1.3.1 Canteen Stick

- Bump `contra.emitter` to 1.0.1

# 1.3.0 Cat Soup

- Remove `contra` in favor of leaner `contra.emitter`

# 1.2.9 Soup Can

- Default to an empty JSON body

# 1.2.2 Chicken Wing

- Events are now properly emitter downstream starting on the top-most layer

# 1.2.0 Kitchen Sink

- Added `.all` to list all measly layers

# 1.1.0 Dreadful Pirate

- Added ability to see all measly layers on the topmost object

# 1.0.3 IPO

- Initial Public Release


gulpfile.js
'use strict';

var prettyBytes = require('pretty-bytes');
var gzipSize = require('gzip-size');
var fs = require('fs');
var path = require('path');
var gulp = require('gulp');
var bump = require('gulp-bump');
var git = require('gulp-git');
var clean = require('gulp-clean');
var rename = require('gulp-rename');
var header = require('gulp-header');
var uglify = require('gulp-uglify');
var browserify = require('browserify');
var streamify = require('gulp-streamify');
var source = require('vinyl-source-stream');
var size = require('gulp-size');

var extended = [
  '/**',
  ' * <%= pkg.name %> - <%= pkg.description %>',
  ' * @version v<%= pkg.version %>',
  ' * @link <%= pkg.homepage %>',
  ' * @license <%= pkg.license %>',
  ' */',
  ''
].join('\n');

var succint = '// <%= pkg.name %>@v<%= pkg.version %>, <%= pkg.license %> licensed. <%= pkg.homepage %>\n';

function build () {
  var pkg = require('./package.json');

  return browserify('./src/measly.js', { debug: true, standalone: 'measly' })
    .bundle()
    .pipe(source('measly.js'))
    .pipe(streamify(header(extended, { pkg : pkg } )))
    .pipe(gulp.dest('./dist'))
    .pipe(streamify(rename('measly.min.js')))
    .pipe(streamify(uglify()))
    .pipe(streamify(header(succint, { pkg : pkg } )))
    .pipe(streamify(size()))
    .pipe(gulp.dest('./dist'));
}

function bumpOnly () {
  var bumpType = process.env.BUMP || 'patch'; // major.minor.patch

  return gulp.src(['./package.json', './bower.json'])
    .pipe(bump({ type: bumpType }))
    .pipe(gulp.dest('./'));
}

function tag () {
  var pkg = require('./package.json');
  var v = 'v' + pkg.version;
  var message = 'Release ' + v;

  return gulp.src('./')
    .pipe(git.commit(message))
    .pipe(git.tag(v, message))
    .pipe(git.push('origin', 'master', '--tags'))
    .pipe(gulp.dest('./'));
}

function publish (done) {
  require('child_process').exec('npm publish', { stdio: 'inherit' }, done);
}

gulp.task('clean', function () {
  gulp.src('./dist', { read: false })
    .pipe(clean());
});

gulp.task('build', ['clean'], build);
gulp.task('bump', bumpOnly);
gulp.task('bump-build', ['bump'], build);
gulp.task('tag', ['bump-build'], tag);
gulp.task('npm', publish);
gulp.task('release', ['tag'], function () {
  gulp.start('npm');
});


sample/sample.js
void function () {
  'use strict';

  var i = 1;
  var addButton = document.querySelector('.cm-add');
  var many = document.querySelector('.ms-many');
  var abortAllButton = document.querySelector('.cm-abort-all');
  var preventButton = document.querySelector('.cm-prevent-all');

  document.body.addEventListener('click', delegateRequest);

  addButton.addEventListener('click', addSpace);
  abortAllButton.addEventListener('click', abortAll);
  preventButton.addEventListener('click', preventAll);

  function tooManySpaces () {
    many.classList.add('ms-show');

    setTimeout(function () {
      many.classList.remove('ms-show');
    }, 2000);
  }

  function addSpace () {
    if (i > 3) {
      tooManySpaces();
      return;
    }
    var header = document.createElement('h3');
    var space = document.createElement('div');
    var requestButton = document.createElement('button');
    var description = document.createElement('pre');
    var layer = measly.thinner({ context: space });
    header.innerText = 'Space #' + i++;
    requestButton.classList.add('cm-request');
    requestButton.innerText = 'Start Request!';
    space.classList.add('ms-space');
    space.appendChild(header);
    space.appendChild(requestButton);
    space.appendChild(description);
    document.body.appendChild(space);

    updateDescription();

    layer.on('create', updateDescription);
    layer.on('always', updateDescription);

    function updateDescription () {
      if (layer.requests.length === 0) {
        description.innerText = 'No requests going on!';
        return;
      }
      var html = '';
      layer.requests.forEach(function (req, i) {
        var cls = '';
        var state;
        if (req.prevented) {
          cls = 'rq-prevented';
          state = 'prevented';
        } else if (req.aborted) {
          cls = 'rq-aborted';
          state = 'aborted';
        } else if (req.cached) {
          cls = 'rq-cached';
          state = 'cached';
        } else if (req.error) {
          cls = 'rq-error';
          state = 'error';
        } else if (req.done) {
          cls = 'rq-done';
          state = 'done';
        } else if (req.requested) {
          cls = 'rq-requested';
          state = 'requested';
        } else {
          state = 'created';
        }
        html += ['<div class="',cls,'">',req.method,', ',req.duration,'s (',state,') </div>\n'].join('');
      });
      description.innerHTML = html;
    }
  }

  function delegateRequest (e) {
    if (!e.target.classList.contains('cm-request')) {
      return;
    }
    var layer = measly.find(e.target);
    var duration = Math.floor(Math.random() * 30 + 2);
    var req = layer.get('http://www.filltext.com/?rows=10&f={firstName}&delay=' + duration);
    req.duration = duration;
  }

  function preventAll () {
    function shield (req) {
      req.prevent();
    }
    measly.on('create', shield);
    preventButton.innerText = 'Shield On..';
    preventButton.classList.add('cm-prevent-on');

    setTimeout(function () {
      measly.off('create', shield);
      preventButton.innerText = 'Prevent for 5s';
      preventButton.classList.remove('cm-prevent-on');
    }, 5000);
  }

  function abortAll () {
    measly.abort();
  }
}();


src/measly.js
'use strict';

var raf = require('raf');
var xhr = require('xhr');
var emitter = require('contra/emitter');
var cache = require('./cache');
var aggregate = require('./aggregate');
var emitCascade = require('./emitCascade');
var methods = ['get', 'post', 'put', 'delete', 'patch'];
var core;
var all = [];

function measly (measlyOptions, parent) {
  var layer = find(measlyOptions.context, true);
  if (layer) {
    return layer;
  }

  layer = emitter({
    layer: thinner,
    parent: parent,
    context: measlyOptions.context,
    children: [],
    requests: [],
    cache: {},
    abort: abort,
    request: request
  }, { throws: false });

  methods.forEach(function addMethod (method) {
    layer[method] = fire.bind(null, method);
  });

  all.push(layer);

  function request (url, opt) {
    var method = opt.method;
    if (opt && opt.xhr && opt.xhr.method) {
      method = opt.xhr.method;
    }
    if (!method) {
      throw new Error('A request method must be specified.');
    }
    return fire(method, url, opt);
  }

  function thinner (opt) {
    var existing = find(opt.context || measlyOptions.context, true);
    if (existing) {
      return existing;
    }
    var child = measly(opt || measlyOptions, layer);
    layer.children.push(child);
    return child;
  }

  function fire (method, url, opt) {
    var fireOptions = opt || {};
    fireOptions.json = 'json' in fireOptions ? fireOptions.json : {};
    fireOptions.url = url;
    fireOptions.method = method.toUpperCase();

    var req = emitter({
      done: false,
      requested: false,
      prevented: false,
      prevent: prevent,
      layer: layer,
      context: fireOptions.context || measlyOptions.context,
      cache: fireOptions.cache || measlyOptions.cache,
      url: url,
      method: method
    }, { throws: false });
    req.abort = abortRequest.bind(null, req);

    raf(go);

    function go () {
      emitCascade(req, 'create', req);
      request();
    }

    function prevent (err, body, statusCode) {
      if (req.prevented) {
        return;
      }
      if (req.requested === true) {
        throw new Error('A request has already been made. Prevent synchronously!');
      }
      req.prevented = true;
      raf(prevented);

      function prevented () {
        var xhr = {
          body: body,
          statusCode: statusCode || err ? 500 : 200
        };
        emitCascade(req, 'cache', err, body);
        done(err, xhr, body);
      }
    }

    function cacheHit () {
      var entry = cache.find(url, layer);
      if (entry) {
        entry.cached = true;
        req.xhr = entry;
        req.prevented = true;
        emitCascade(req, 'cache', entry.error, entry.body);
        done(entry.error, entry, entry.body);
      }
    }

    function request () {
      if (method === 'GET') {
        cacheHit();
      }
      if (req.prevented) {
        return;
      }
      req.requested = true;
      req.xhr = xhr(fireOptions, done);
      emitCascade(req, 'request', req.xhr);
    }

    function done (err, res, body) {
      req.error = err;
      req.response = body;
      req.done = true;
      if (req.cache && !res.cached) {
        layer.cache[url] = {
          expires: cache.expires(req.cache),
          error: err,
          body: body,
          statusCode: res.statusCode
        };
      }
      if (err) {
        emitCascade(req, 'error', err, body);
        emitCascade(req, err.statusCode, err, body);
      } else {
        emitCascade(req, 'data', body);
      }
      untrack(req);
    }

    track(req);
    return req;
  }

  function abort () {
    aggregate.requests(layer, true).forEach(abortRequest);
  }

  function abortRequest (req) {
    req.aborted = true;
    req.prevented = true;
    emitCascade(req, 'abort', req.xhr);

    if (req.xhr) {
      req.xhr.abort();
    }
    untrack(req);
  }

  function track (req) {
    layer.requests.push(req);
  }

  function untrack (req) {
    var i = layer.requests.indexOf(req);
    var spliced = layer.requests.splice(i, 1);
    if (spliced.length) {
      emitCascade(req, 'always', req.error, req.response, req);
    }
  }

  return layer;
}

function find (context, shallow) {
  if (core === void 0) {
    return;
  }
  var depleted;
  var layers = aggregate.contexts(core);
  while (context && !depleted) {
    var needle = sameContext(layers);
    if (needle) {
      return needle.layer;
    }
    context = context.parentNode;
    depleted = shallow === true;
  }
  function sameContext (layers) {
    var i;
    var len = layers.length;
    for (i = 0; i < len; i++) {
      var layer = layers[i];
      if (layer.context === context) {
        return layer;
      }
    }
  }
}

module.exports = core = measly({
  context: global.document.body,
  base: ''
});

core.find = find;
core.all = all;
